service/src/main/java/bio/terra/workspace/service/workspace/flight/DeleteControlledSamResourcesStep.java

```diff
@@ -0,0 +1,82 @@
+package bio.terra.workspace.service.workspace.flight;
+
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.FlightMap;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.service.iam.AuthenticatedUserRequest;
+import bio.terra.workspace.service.iam.SamService;
+import bio.terra.workspace.service.resource.controlled.ControlledResource;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ControlledResourceKeys;
+import java.util.List;
+
+/**
+ * A step which deletes the Sam data of the controlled resources provided by {@link
+ * ListControlledResourceIdsStep} Multiple deletions are run in parallel as resources are
+ * independent.
+ */
+public class DeleteControlledSamResourcesStep implements Step {
+
+  private final SamService samService;
+  private final AuthenticatedUserRequest userRequest;
+
+  public DeleteControlledSamResourcesStep(
+      SamService samService, AuthenticatedUserRequest userRequest) {
+    this.samService = samService;
+    this.userRequest = userRequest;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    FlightMap workingMap = flightContext.getWorkingMap();
+
+    @SuppressWarnings("unchecked")
+    List<ControlledResource> controlledResourceList =
+        workingMap.get(ControlledResourceKeys.CONTROLLED_RESOURCE_LIST, List.class);
+
+    try {
```

Comment:
Reviewers:
This implementation is kind of hacky, but I think we need to run the Sam requests in parallel to avoid too long on this request. I'm not super familiar with streams - is there a cleaner/more canonical pattern for this?

Reply:
Is this a premature optimization? My theory is that the typical number of controlled resources in a workspace will be maybe 10. So do we really have a performance problem here?

That said, this doesn't strike me as hacky. Seems like a perfect case for parallelStream() and the code is straightforward. My only concern us that Sam will not welcome the rapid fire requests. I'm not sure how things are queued or how they apply backpressure. It would be counterproductive if the result of parallelism here is failure/retries in the SamService. Might be worth checking with CI team.

Reply:
I was working on the assumption that an average workspace might have 10s of resources, and that ~30 resources would not be unreasonable. This comment made check the traces on `SamService.deleteControlledResource` though, and it turns out I was overestimating how long the Sam calls took by 3-4x (each one is actually ~250ms in the traces I sampled). something something measure twice...

Removed the parallelism for now. If this becomes an issue in the future there's room for optimization, including discussion with CI about how Sam wants to handle many requests.